.github/workflows/build_and_test_dbt.yaml

```diff
@@ -0,0 +1,87 @@
+name: build-and-test-dbt
+
+on:
+  pull_request:
+    branches: [master]
+  push:
+    branches: [master]
+
+jobs:
+  build-and-test-dbt:
+    runs-on: ubuntu-latest
+    # These permissions are needed to interact with GitHub's OIDC Token endpoint
+    # so that we can authenticate with AWS
+    permissions:
+      id-token: write
+      contents: read
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Install dbt requirements
+        uses: ./.github/actions/install_dbt_requirements
+
+      - name: Load environment variables
+        uses: ./.github/actions/load_environment_variables
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v2
+        with:
+          role-to-assume: ${{ secrets.AWS_IAM_ROLE_TO_ASSUME_ARN }}
+          aws-region: us-east-1
+
+      - name: Configure dbt environment
+        uses: ./.github/actions/configure_dbt_environment
+
+      - name: Cache dbt state directory
+        id: cache
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.PROJECT_DIR }}/${{ env.STATE_DIR }}
+          key: ${{ env.CACHE_NAME }}-${{ env.CACHE_KEY }}
+          restore-keys: |
+            ${{ env.CACHE_NAME }}-master
+
+      - if: ${{ steps.cache.outputs.cache-hit == 'true' }}
+        name: Set command args to build/test modified resources
+        run: echo "MODIFIED_RESOURCES_ONLY=true" >> "$GITHUB_ENV"
+        shell: bash
+
+      - if: ${{ steps.cache.outputs.cache-hit != 'true' }}
+        name: Set command args to build/test all resources
+        run: echo "MODIFIED_RESOURCES_ONLY=false" >> "$GITHUB_ENV"
+        shell: bash
+
+      - name: Test dbt macros
+        run: dbt run-operation test_all
+        working-directory: ${{ env.PROJECT_DIR }}
+        shell: bash
+
+      - name: Build models
+        run: |
+          if [[ $MODIFIED_RESOURCES_ONLY == 'true' ]]; then
+            echo "Running build on modified/new resources only"
+            dbt run -t "$TARGET" -s state:modified state:new --defer --state "$STATE_DIR"
+          else
+            echo "Running build on all resources"
+            dbt run -t "$TARGET"
+          fi
+        working-directory: ${{ env.PROJECT_DIR }}
+        shell: bash
+
+      - name: Test models
+        run: |
+          if [[ $MODIFIED_RESOURCES_ONLY == 'true' ]]; then
+            echo "Running tests on modified/new resources only"
+            dbt test -t "$TARGET" -s state:modified state:new --state "$STATE_DIR"
```

Comment:
**question (non-blocking):** So dbt knows to run tests for any part of the DAG where the models are modified/new. Does it also know to run tests for anywhere that the _tests themselves_ are modified/new? I.e. if we start creating PRs that don't modify a view but add many tests, will dbt know to run them?

Reply:
> Does it also know to run tests for anywhere that the tests themselves are modified/new?

That's my understanding, yes! Tests are a type of node, just like models, and my read of [the docs](https://docs.getdbt.com/reference/node-selection/methods#the-state-method) is that state selection applies to all node types. But I haven't confirmed this locally yet, I'll do so now!